### PR TITLE
Refactor features, implement new design for navigation & user menu

### PR DESCRIPTION
**Pull request by ryanclark**
This PR has a few architecture changes around features/routes, and implements the new navigation and user menu designs.

### Designs

Navigation - https://www.figma.com/file/v6GunK50D2VC7w7I2FBDNf/Teleport-Access-Management?node-id=121%3A6710&t=NDvvGzZNMMhw1oCp-0
User Menu - https://www.figma.com/file/Gpjs9vjhzUKF1GDbeG9JGE/Application-Design-System?node-id=997%3A2813&t=LE6vO4JJH2R3QCQu-0

### SVG Icons

I've added a bunch of the icons we're using in the new designs as SVG React components. These work way nicer than using font awesome (they're font awesome SVGs, but not using text for rendering). They look much more crisp than using the font alternative.

We should probably revisit the new `SVGIcons` components and include all of them at some point, but this PR is a good start to using them. They're named after their usage instead of their font awesome name, so it's more obvious which icon is to be used where.

### FeatureContext

The feature context seemed to only exist to pass through the features from `Teleport`/`TeleportE` through to the `Main` component, where it would then render the routes and register the features inside the monolithic `TeleportContext`.

I've changed this so the feature context only contains the features that are actually available to the authenticated user, and we can now use `useFeatures()` everywhere instead of pulling in `TeleportContext` and looking at `ctx.features`.

Features also now have a slightly different shape, as they define where they live in the new navigation (either under "Resources" or "Management", or not at all).

### Cleanup of `useX()`

This also cleans up a few non-helpful hook abstractions by putting their hooks into the main component that calls them. This simplifies `TopNav`, `Main` and `Clusters`.

It also means we can get rid of a few `ReturnType` values being used as the state of components. This is a major annoyance when trying to track down where different state values come from and the origins of props. We should strive to move away from this pattern.

## UI Changes

Here's a short video showing the new animations to go with the design

https://user-images.githubusercontent.com/7922109/214074373-b7a1ff48-5f04-4670-a6d3-dec45b9828ed.mov

### New navigation

image

image

image

image

### User menu

image

### Other UX changes

The top bar is now 72px, matching the height of the design. I've overridden a few styles in our DarkSelect component to get the cluster selector inline with the designs also.

image

### CODE REVIEW

**Review by kimlisa — COMMENTED** [review]
i really love the new navigation and had fun clicking around. couple of other things we need to do:

- add license headers to new files
- the `username` dropdown is also missing a hover highlight like we do for everything else 

optional nits:
- maybe it's just me that is bothered by this, but when I click on the `username` drop down, the animation seems different from the `management/resource` drop down, it seems to take longer and text slides
- prior we supported `keyboard tab` highlighting on the side navs, but we are missing it on the new nav (not that we had great support for it before so more like fyi)

but otherwise this is awesome stuff!

File: web/packages/design/src/SVGIcon/AccessRequests.tsx [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      width={size}
+      height={size}
+      fill={fill}
+    >
+      <path d="M7.5 7.875V5.625C7.5 5.00156 6.99844 4.5 6.375 4.5H4.5V0.375C4.5 0.16875 4.33125 0 4.125 0H3.375C3.16875 0 3 0.16875 3 0.375V4.5H1.125C0.501562 4.5 0 5.00156 0 5.625V7.875C0 8.49844 0.501562 9 1.125 9H3V23.625C3 23.8312 3.16875 24 3.375 24H4.125C4.33125 24 4.5 23.8312 4.5 23.625V9H6.375C6.99844 9 7.5 8.49844 7.5 7.875ZM6 7.5H1.5V6H6V7.5ZM13.125 15H11.25V0.375C11.25 0.16875 11.0813 0 10.875 0H10.125C9.91875 0 9.75 0.16875 9.75 0.375V15H7.875C7.25156 15 6.75 15.5016 6.75 16.125V18.375C6.75 18.9984 7.25156 19.5 7.875 19.5H9.75V23.625C9.75 23.8312 9.91875 24 10.125 24H10.875C11.0813 24 11.25 23.8312 11.25 23.625V19.5H13.125C13.7484 19.5 14.25 18.9984 14.25 18.375V16.125C14.25 15.5016 13.7484 15 13.125 15ZM12.75 18H8.25V16.5H12.75V18ZM19.875 7.5H18V0.375C18 0.16875 17.8312 0 17.625 0H16.875C16.6688 0 16.5 0.16875 16.5 0.375V7.5H14.625C14.0016 7.5 13.5 8.00156 13.5 8.625V10.875C13.5 11.4984 14.0016 12 14.625 12H16.5V23.625C16.5 23.8312 16.6688 24 16.875 24H17.625C17.8312 24 18 23.8312 18 23.625V12H19.875C20.4984 12 21 11.4984 21 10.875V8.625C21 8.00156 20.4984 7.5 19.875 7.5ZM19.5 10.5H15V9H19.5V10.5Z" />
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/design/src/SVGIcon/AccessRequests.tsx:17]: From font awesome, I copied the SVG from Figma

File: web/packages/teleport/src/features.tsx [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+
+  hasAccess(flags: FeatureFlags) {
+    return flags.nodes;
+  }
+}
+
+export class FeatureApps implements TeleportFeature {
+  category = NavigationCategory.Resources;
+
+  topNavTitle = 'Applications';
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/features.tsx:127]: For some reason we have topNavTitle that goes where the cluster selector is when browsing a non cluster specific URL.

We could probably just use `route.title` though right? And then we can get rid of `topNavTitle`.

File: web/packages/teleport/src/Recordings/__snapshots__/Recordings.story.test.tsx.snap [hunk 3]
```diff
@@ -254,63 +254,61 @@ exports[`rendering of Session Recordings 1`] = `
 }
 
 .c3 .react-select__value-container {
-  padding: 0 8px;
+  padding: 0 16px;
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/Recordings/__snapshots__/Recordings.story.test.tsx.snap:257]: I've scoped the changes to only the cluster selector at the top, so this is fixed

File: web/packages/teleport/src/Discover/Discover.tsx [hunk 4]
```diff
@@ -81,85 +57,23 @@ function DiscoverContent() {
...
-      onBannerDismiss={dismissAlert}
-    >
-      <MainContainer>
-        <Prompt
-          message={nextLocation => {
-            if (nextLocation.pathname === cfg.routes.discover) return true;
-            return 'Are you sure you want to exit the “Add New Resource” workflow? You’ll have to start from the beginning next time.';
-          }}
-          when={selectedResource.shouldPrompt(currentStep)}
-        />
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/Discover/Discover.tsx]: Added it back, we can discuss the wording in our sync tomorrow

File: web/packages/teleport/src/Discover/Navigation/StepItem.tsx [hunk 5]
```diff
@@ -114,9 +107,26 @@ const CheckedBullet = styled(Bullet)`
   }
 `;
 
-const StepsContainer = styled<{ active: boolean }>(Text)`
+const StepsContainer = styled.div<{ active: boolean }>`
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/Discover/Navigation/StepItem.tsx:110]: Fixed, nice spot

**Comment by ryanclark** [comment]
> i really love the new navigation and had fun clicking around. couple of other things we need to do:

Cheers!

> * add license headers to new files

Done

> * the `username` dropdown is also missing a hover highlight like we do for everything else

The design doesn't have a hover state & Kenny was happy with just the text colour change, but I'll re-confirm with him

> optional nits:
> 
> * maybe it's just me that is bothered by this, but when I click on the `username` drop down, the animation seems different from the `management/resource` drop down, it seems to take longer and text slides

Yeah this is by design for actual popovers, the text also slides on the dropdown for resources but is a smaller change as it's coming from top down instead of right to left

> * prior we supported `keyboard tab` highlighting on the side navs, but we are missing it on the new nav (not that we had great support for it before so more like fyi)

I've added keyboard support - both tab and pressing the up/right arrows will navigate between links. Pressing right/left will open/close the Access Request dropdown when that's selected. I've also added keyboard navigation to the navigation selector between management/resources.
> but otherwise this is awesome stuff!

**Review by gzdunek — COMMENTED** [review]
I added a few comments, but it's hard to review so many changes :(

I'm quite worried about the styles, it seems like values for font sizes, margins, colors, shadows are hardcoded in the components (instead of adding/taking them from `theme`).

Overall, good job!

**Comment by ryanclark** [comment]
> I'm quite worried about the styles, it seems like values for font sizes, margins, colors, shadows are hardcoded in the components (instead of adding/taking them from `theme`).

Same, but our theme is so inconsistent that I figured we'll wait for the design system changes next quarter. I searched the codebase for some of the colours used from the design, and they only existed as hard-coded elsewhere.

File: web/packages/teleport/src/Clusters/Clusters.tsx [hunk 6]
```diff
@@ -14,42 +14,54 @@ See the License for the specific language governing permissions and
...
+export function Clusters() {
   const ctx = useTeleport();
-  const state = useClusters(ctx);
-  return <Clusters {...state} />;
-}
 
-export function Clusters(props: ReturnType<typeof useClusters>) {
-  const { clusters, enabledFeatures, initAttempt } = props;
+  const [clusters, setClusters] = useState([]);
+  const { attempt, run } = useAttempt();
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/Clusters/Clusters.tsx:39]: I just moved the stuff from `useCluster()` into this component, I didn't really want to refactor it completely - only the parts that previously used `ctx.features`.

File: web/packages/teleport/src/Discover/Navigation/StepList.tsx [hunk 7]
```diff
@@ -16,25 +16,32 @@
...
+      <StepItem
+        key={index}
+        view={view}
+        currentStep={props.currentStep}
+        index={index}
+      />
+    );
+
+    index += 1;
+  }
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/Discover/Navigation/StepList.tsx:44]: Yeah I'm more a fan of `for .. of` loops to be honest, but I agree with the rename to `startIndex`

File: web/packages/teleport/src/Main/Main.tsx [hunk 8]
```diff
@@ -14,48 +14,77 @@ See the License for the specific language governing permissions and
...
-      initialAlerts: props.initialAlerts,
-      customBanners: props.customBanners,
-    });
+  const ctx = useTeleport();
+  const history = useHistory();
+
+  const { attempt, setAttempt, run } = useAttempt('processing');
+
+  useEffect(() => {
+    if (ctx.storeUser.state) {
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/Main/Main.tsx:65]: Not sure, this was the original code in `useMain` that I just moved into this function

File: web/packages/teleport/src/Main/Main.tsx [hunk 9]
```diff
@@ -14,48 +14,77 @@ See the License for the specific language governing permissions and
...
+    }
+
+    run(() => ctx.init());
+  }, []);
+
+  const featureFlags = ctx.getFeatureFlags();
+
+  const features = useMemo(
+    () => props.features.filter(feature => feature.hasAccess(featureFlags)),
+    [featureFlags]
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/Main/Main.tsx:77]: Good shout

File: web/packages/teleport/src/components/UserMenuNav/UserMenuNav.tsx [hunk 10]
```diff
@@ -14,190 +14,238 @@
...
+  const ref = useRef<HTMLDivElement>();
+
+  const ctx = useTeleport();
+  const clusterId = ctx.storeUser.getClusterId();
+  const features = useFeatures();
+
+  const initial =
+    username && username.length ? username.trim().charAt(0).toUpperCase() : '';
+
+  function handleClickOutside(event) {
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/components/UserMenuNav/UserMenuNav.tsx]: Moved & added event

**Comment by ryanclark** [comment]
> > maybe it's just me that is bothered by this, but when I click on the username drop down, the animation seems different from the management/resource drop down, it seems to take longer and text slides
> 
> I agree, I'd even say that animations for dropdowns are not needed at all. For me they are quite distracting :) I checked and GitHub/Slack/native dropdowns are not animated.

I guess it depends on where you look. Google Calendar/Material UI use animations to show dropdowns. You can check out our conversations regarding the animations & their fine tuning in #new-navigation in Slack

File: web/packages/teleport/src/components/UserMenuNav/UserMenuNav.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/components/UserMenuNav/UserMenuNav.tsx]: Moved it back as ESLint complains, but I've added it as a dependency

[the remark this answers is not in the document]
Comment (gzdunek) [reply web/packages/teleport/src/components/UserMenuNav/UserMenuNav.tsx]: Hmm why is ESLint complaining? 
We already have functions in `useEffect` like here https://github.com/gravitational/teleport/blob/b23dcd680a07631f7935dc0a469d189b214977cc/web/packages/shared/hooks/useEscape.ts#L21

[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/components/UserMenuNav/UserMenuNav.tsx]:
```
  119:7  error  Move function declaration to function body root  no-inner-declarations
```

I guess ESLint doesn't recognise anonymous lambdas as function declarations, but I'm not a fan of defining functions as constant lambas - I'd rather just go for a named function block.

File: web/packages/teleport/src/Navigation/NavigationSwitcher.tsx [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+export function NavigationSwitcher(props: NavigationSwitcherProps) {
+  const [open, setOpen] = useState(false);
+
+  const ref = useRef<HTMLDivElement>();
+  const activeValueRef = useRef<HTMLDivElement>();
+  const firstValueRef = useRef<HTMLDivElement>();
+
+  const activeItem = props.items.find(item => item === props.value);
+
+  function handleClickOutside(event: MouseEvent) {
```
[the remark this answers is not in the document]
Comment (ryanclark) [reply web/packages/teleport/src/Navigation/NavigationSwitcher.tsx:116]: Good shout

**Review by strideynet — APPROVED** [review]

**Review by gzdunek — APPROVED** [review]

**Review by kimlisa — APPROVED** [review]

**Pull request merged**